crates/ruff_benchmark/benches/ty.rs

```diff
@@ -347,10 +348,141 @@ fn benchmark_many_tuple_assignments(criterion: &mut Criterion) {
     });
 }
 
+struct ProjectBenchmark<'a> {
+    project: SetupProject<'a>,
+    fs: MemoryFileSystem,
+    max_diagnostics: usize,
+}
+
+impl<'a> ProjectBenchmark<'a> {
+    fn new(project: RealWorldProject<'a>, max_diagnostics: usize) -> Self {
+        let setup_project = project.setup().expect("Failed to setup project");
+        let fs = setup_project
+            .copy_to_memory_fs()
+            .expect("Failed to copy project to memory fs");
+
+        Self {
+            project: setup_project,
+            fs,
+            max_diagnostics,
+        }
+    }
+
+    fn setup_iteration(&self) -> ProjectDatabase {
+        let system = TestSystem::new(InMemorySystem::from_memory_fs(self.fs.clone()));
+
+        let src_root = SystemPath::new("/");
+        let mut metadata = ProjectMetadata::discover(src_root, &system).unwrap();
+
+        metadata.apply_options(Options {
+            environment: Some(EnvironmentOptions {
+                python_version: Some(RangedValue::cli(self.project.config.python_version)),
+                python: (!self.project.config().dependencies.is_empty())
+                    .then_some(RelativePathBuf::cli(SystemPath::new(".venv"))),
+                ..EnvironmentOptions::default()
+            }),
+            ..Options::default()
+        });
+
+        let mut db = ProjectDatabase::new(metadata, system).unwrap();
+
+        db.project().set_included_paths(
+            &mut db,
+            self.project
+                .check_paths()
+                .iter()
+                .map(|path| path.to_path_buf())
+                .collect(),
+        );
+
+        db
+    }
+}
+
+#[track_caller]
+fn bench_project(benchmark: &ProjectBenchmark, criterion: &mut Criterion) {
+    fn check_project(db: &mut ProjectDatabase, max_diagnostics: usize) {
+        let result = db.check();
+        let diagnostics = result.len();
+
+        assert!(
+            diagnostics > 1 && diagnostics <= max_diagnostics,
+            "Expected between {} and {} diagnostics but got {}",
+            1,
+            max_diagnostics,
+            diagnostics
+        );
```

Comment:
Why do we have this? As some kind of low-fidelity filter that we do "actual type checking work" on the project? I'm not opposed to it, but it might require updating the thresholds from time to time?